cupy/cuda/memory.pyx

```diff
@@ -675,26 +681,46 @@ cdef object _get_chunk(SingleDeviceMemoryPool pool, size_t size,
 
 
 cdef BaseMemory _try_malloc(SingleDeviceMemoryPool pool, size_t size):
+    # Be optimistic to avoid locking the entire malloc section.
+    total = pool._total_bytes + size
+    if pool._total_bytes_limit != 0 and pool._total_bytes_limit < total:
+        raise OutOfMemoryError(size, total, pool._total_bytes_limit)
+
+    mem = None
     try:
-        return pool._alloc(size).mem
+        mem = pool._alloc(size).mem
     except runtime.CUDARuntimeError as e:
         if e.status != runtime.errorMemoryAllocation:
             raise
         pool.free_all_blocks()
         try:
-            return pool._alloc(size).mem
+            mem = pool._alloc(size).mem
         except runtime.CUDARuntimeError as e:
             if e.status != runtime.errorMemoryAllocation:
                 raise
             gc.collect()
             pool.free_all_blocks()
             try:
-                return pool._alloc(size).mem
+                mem = pool._alloc(size).mem
             except runtime.CUDARuntimeError as e:
                 if e.status != runtime.errorMemoryAllocation:
                     raise
-    total = size + pool.total_bytes()
-    raise OutOfMemoryError(size, total)
+
+    if mem is None:
+        raise OutOfMemoryError(size, total)
+
+    rlock.lock_fastrlock(pool._total_bytes_lock, -1, True)
+    try:
+        total = pool._total_bytes + size
+        if pool._total_bytes_limit != 0 and pool._total_bytes_limit < total:
+            del mem
```

Comment:
This intends to make it clear that `mem` should be released immediately.

Reply:
I don't think `del` should be used for that purpose.
It actually has completely different semantics. It's a common misconception that it would release the memory and we shouldn't encourage it.

Reply:
Ok, I removed it.